### ISSUE TITLE
What the overall impact on Rails in case of using this gem?

### ISSUE THREAD

**Issue by gogainda** [danielpclark/faster_path#1]

**Comment by danielpclark** [comment]
This gem will have more path related methods rewritten to continue speeding up performance.  It's only 1 day old at the moment.  But here's the current stack profile change.

# before

```
       user     system      total        real
100 requests 29.670000   1.690000  31.360000 ( 31.567261)
Running `stackprof tmp/2016-06-10T11:32:34-04:00-stackprof-cpu-myapp.dump`. Execute `stackprof --help` for more info
==================================
  Mode: cpu(1000)
  Samples: 7732 (1.64% miss rate)
  GC: 1067 (13.80%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
      1831  (23.7%)        1831  (23.7%)     Pathname#chop_basename
      1034  (13.4%)         880  (11.4%)     Hike::Index#entries
      1382  (17.9%)         456   (5.9%)     BetterErrors::ExceptionExtension#set_backtrace
       374   (4.8%)         370   (4.8%)     Sprockets::Mime#mime_types
      1387  (17.9%)         322   (4.2%)     Pathname#plus
      1419  (18.4%)         284   (3.7%)     BindingOfCaller::BindingExtensions#callers
       381   (4.9%)         193   (2.5%)     Hike::Index#find_aliases_for
      1574  (20.4%)         187   (2.4%)     Pathname#+
       375   (4.8%)         185   (2.4%)     Hike::Index#sort_matches
       260   (3.4%)         130   (1.7%)     ActionView::PathResolver#find_template_paths
       127   (1.6%)         108   (1.4%)     Hike::Index#build_pattern_for
      1655  (21.4%)         100   (1.3%)     Hike::Index#match
       207   (2.7%)          95   (1.2%)     Hike::Index#pattern_for
        74   (1.0%)          74   (1.0%)     Hike::Trail#stat
      2395  (31.0%)          68   (0.9%)     Pathname#join
        62   (0.8%)          62   (0.8%)     Sass::SCSS::Parser#tok
       117   (1.5%)          62   (0.8%)     Hike::Index#initialize
        48   (0.6%)          48   (0.6%)     Time#compare_with_coercion
        50   (0.6%)          47   (0.6%)     ActiveSupport::FileUpdateChecker#watched
        41   (0.5%)          41   (0.5%)     Sass::Script::Lexer#scan
      5175  (66.9%)          38   (0.5%)     Sprockets::Rails::Helper#check_errors_for
        37   (0.5%)          37   (0.5%)     Sass::Tree::Visitors::Base#node_name
       129   (1.7%)          29   (0.4%)     ActiveSupport::FileUpdateChecker#max_mtime
       771  (10.0%)          28   (0.4%)     Pathname#relative?
        26   (0.3%)          26   (0.3%)     ActiveSupport::SafeBuffer#initialize
       101   (1.3%)          25   (0.3%)     Sprockets::Caching#cache_set
        96   (1.2%)          24   (0.3%)     Sass::Importers::Filesystem#find_real_file
        98   (1.3%)          23   (0.3%)     Sprockets::Asset#dependency_fresh?
        38   (0.5%)          21   (0.3%)     ActiveSupport::Inflector#camelize
      7830 (101.3%)          21   (0.3%)     Hike::Index#find_in_paths
```
# after

```
       user     system      total        real
100 requests 25.990000   1.600000  27.590000 ( 27.773812)
Running `stackprof tmp/2016-06-10T11:59:13-04:00-stackprof-cpu-myapp.dump`. Execute `stackprof --help` for more info
==================================
  Mode: cpu(1000)
  Samples: 6931 (0.04% miss rate)
  GC: 923 (13.32%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
      1786  (25.8%)        1786  (25.8%)     Pathname#chop_basename
      1073  (15.5%)         933  (13.5%)     Hike::Index#entries
      1207  (17.4%)         401   (5.8%)     BetterErrors::ExceptionExtension#set_backtrace
       401   (5.8%)         400   (5.8%)     Sprockets::Mime#mime_types
      1390  (20.1%)         290   (4.2%)     Pathname#plus
      1219  (17.6%)         254   (3.7%)     BindingOfCaller::BindingExtensions#callers
      1596  (23.0%)         206   (3.0%)     Pathname#+
       337   (4.9%)         173   (2.5%)     Hike::Index#find_aliases_for
       340   (4.9%)         170   (2.5%)     Hike::Index#sort_matches
       258   (3.7%)         129   (1.9%)     ActionView::PathResolver#find_template_paths
       112   (1.6%)          98   (1.4%)     Hike::Index#build_pattern_for
      1647  (23.8%)          94   (1.4%)     Hike::Index#match
       193   (2.8%)          92   (1.3%)     Hike::Index#pattern_for
      2398  (34.6%)          79   (1.1%)     Pathname#join
        78   (1.1%)          78   (1.1%)     Hike::Trail#stat
        60   (0.9%)          53   (0.8%)     ActiveSupport::FileUpdateChecker#watched
      5198  (75.0%)          47   (0.7%)     Sprockets::Rails::Helper#check_errors_for
       722  (10.4%)          46   (0.7%)     Pathname#relative?
        43   (0.6%)          43   (0.6%)     Time#compare_with_coercion
        65   (0.9%)          35   (0.5%)     Hike::Index#initialize
        48   (0.7%)          30   (0.4%)     Sprockets::Engines#deep_copy_hash
        26   (0.4%)          26   (0.4%)     ActionView::Helpers::AssetUrlHelper#compute_asset_extname
       111   (1.6%)          26   (0.4%)     ActiveSupport::FileUpdateChecker#max_mtime
        49   (0.7%)          23   (0.3%)     ActiveSupport::Inflector#camelize
        22   (0.3%)          22   (0.3%)     ActiveSupport::SafeBuffer#initialize
       100   (1.4%)          21   (0.3%)     Sprockets::Asset#dependency_fresh?
        53   (0.8%)          15   (0.2%)     Rack::MiniProfiler::TimerStruct::Sql#initialize
        15   (0.2%)          15   (0.2%)     ThreadSafe::NonConcurrentCacheBackend#[]
      7854 (113.3%)          15   (0.2%)     Hike::Index#find_in_paths
        17   (0.2%)          15   (0.2%)     block (2 levels) in class_attribute
```

So the changes may seem small at the moment but with 100 requests it looks like it has shaved off 5 seconds total on request time.

Again, more methods will be implemented to further increase performance.

**Comment by gogainda** [comment]
Looks promising, thanks for your work!

**Comment by danielpclark** [comment]
Did the math.  Just that 5 seconds alone is a 15% performance improvement.

**Comment by gogainda** [comment]
Not bad for the first release

**Comment by danielpclark** [comment]
It turns out my test was inaccurate.  This gem previously incorrectly implemented class methods on Pathname with the monkey-patch rather than instance methods.  Now with that fixed in gem release 0.0.9 my rails application has a **66% performance boost!!!**  I'm so stoked!  https://github.com/danielpclark/faster_path/pull/16
